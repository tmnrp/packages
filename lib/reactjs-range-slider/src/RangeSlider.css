* {
  box-sizing: border-box;
  padding: 0;
  margin: 0;
}

.range-slider-container {
  display: inline-flex;
}

.range-slider-container.legend-top {
  flex-direction: column-reverse;
}

.range-slider-container.legend-right {
  flex-direction: row;
}

.range-slider-container.legend-bottom {
  flex-direction: column;
}

.range-slider-container.legend-left {
  flex-direction: row-reverse;
}

.min-wrap,
.max-wrap {
  margin: 5px;
}

.range-slider-wrap {
  display: flex;
  align-items: center;
  background-color: lightblue;
}

.range-slider-wrap.vbox .range-slider {
  -webkit-appearance: slider-vertical;
}

.range-slider-wrap.hbox {
  flex-direction: row;
}
.range-slider-wrap.vbox {
  flex-direction: column-reverse;
}

.legend {
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: lightcoral;
}

.legend.start {
  justify-content: flex-start;
}

.legend.end {
  justify-content: flex-end;
}

.range-slider::-webkit-slider-thumb,
.range-slider::-moz-range-thumb {
  cursor: pointer;
}

.range-slider:hover {
  opacity: 1;
}

.custom-range-slider {
  display: flex;
  justify-content: center;
  align-items: center;
}

.range-slider-value {
  position: absolute;
  font-size: smaller;
}

.range-slider.square,
.range-slider.circle,
.range-slider.virus {
  -webkit-appearance: none !important;
}

.range-slider {
  -webkit-appearance: none;
  opacity: 0.8;
  background: linear-gradient(
    to right,
    #82cfd0 0%,
    #82cfd0 50%,
    #fff 50%,
    #fff 100%
  );

  border: solid 1px #82cfd0;
  outline: none;
  transition: background 450ms ease-in;
}

/*----------------------------------*/
.range-slider.virus {
  background-color: rgb(151, 223, 219);
}

.range-slider.virus::-webkit-slider-thumb {
  -webkit-appearance: none;
  appearance: none;
  width: 25px;
  height: 25px;
  background: url("./virus-icon.png");
  background-repeat: no-repeat;
  background-size: 25px 25px;
  cursor: pointer;
}

/*----------------------------------*/
.range-slider.circle {
  background-color: rgb(243, 174, 174);
}

.range-slider.circle::-webkit-slider-thumb {
  -webkit-appearance: none;
  appearance: none;
  width: 25px;
  height: 25px;
  border-radius: 50%;
  background: red;
  cursor: pointer;
}

/*----------------------------------*/
.range-slider.square .react-rangeslider {
  background-color: lightgrey;
}

.range-slider.square::-webkit-slider-thumb {
  -webkit-appearance: none;
  appearance: none;
  width: 25px;
  height: 25px;
  background: #4caf50;
  cursor: pointer;
}
