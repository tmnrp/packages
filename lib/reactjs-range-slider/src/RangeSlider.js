import "./RangeSlider.css";
import React, { useState } from "react";

export const POSITION_LIST = ["top", "right", "bottom", "left"];
const ALIGNMENT_LIST = ["start", "end"];
const NUMBERS_REG = "^[0-9]*$";

const RangeSlider = (props) => {
  const [value, setValue] = useState(getSliderValue(props.value));
  const legendPositionCls = getLegendPositionCls(props.legendPosition);
  const legend = props.legend && !!legendPositionCls;
  const sliderThumbShapeCls = getSliderThumbShapeCls(props.sliderThumbShape);
  const min = getMin(props.min);
  const max = getMax(props.max, min);
  const step = getStep(props.step);

  //
  return (
    <span
      className={`range-slider-container ${legendPositionCls}`}
      data-testid="range-slider-container-testid"
    >
      <span className={`range-slider-wrap ${props.alignment}`}>
        <span className="min-wrap">{props.min}</span>
        <span className="custom-range-slider">
          {getLegendOnSlider(value, props.legenOnTopOfSlider)}
          <input
            type="range"
            name="rangeslider"
            id="rangeslider"
            className={sliderThumbShapeCls}
            step={step}
            value={value}
            min={min}
            max={max}
            onChange={(e) =>
              onSlide(e.currentTarget, setValue, props.legenOnTopOfSlider)
            }
          />
        </span>
        <span className="max-wrap">{props.max}</span>
      </span>
      {getLegend(
        props.legendAlignment,
        props.legendLabel,
        value,
        props.legendStyle,
        legend
      )}
    </span>
  );
};

export const getLegend = (alignment, label, value, legendStyle, legend) => {
  if (!legend) {
    return null;
  }

  alignment = ALIGNMENT_LIST.includes(alignment)
    ? `legend ${alignment}`
    : "legend";

  return (
    <span
      data-testid="legend-testid"
      className={alignment}
      style={legendStyle}
    >{`${label || ""} ${value || ""}`}</span>
  );
};

export const getLegendOnSlider = (value, legenOnTopOfSlider) => {
  return legenOnTopOfSlider ? (
    <React.Fragment>
      <span className="range-slider-thumb"></span>
      <span className="range-slider-value">{value}</span>
    </React.Fragment>
  ) : null;
};

export const getLegendPositionCls = (legendPosition) => {
  if (!POSITION_LIST.includes(legendPosition)) {
    return false;
  }
  return "legend-" + legendPosition;
};

export const getSliderThumbShapeCls = (sliderThumbShape) => {
  return sliderThumbShape ? `range-slider ${sliderThumbShape}` : "range-slider";
};

export const onSlide = (el, setValue, legenOnTopOfSlider) => {
  const value = el.value;
  setValue(value);

  // Slider gradiant fill
  el.style.background =
    "linear-gradient(to right, #82CFD0 0%, #82CFD0 " +
    value +
    "%, #fff " +
    value +
    "%, white 100%)";

  if (legenOnTopOfSlider) {
    legendOnTopOfSliderHandler(value);
  }
};

export const legendOnTopOfSliderHandler = (value) => {
  const rsv = document.querySelector(".range-slider-value");
  if (value && rsv) {
    const processLeft =
      (value * (131 - 25)) / 100 + value.toString().length * 8;
    rsv.style.left = `${processLeft}px`;
  }
};

export const getSliderValue = (value) => {
  return isInteger(value) ? value : 0;
};

export const getMin = (value) => {
  return isInteger(value) ? value : 0;
};

export const getMax = (max, min) => {
  max = isInteger(max) ? max : 100;
  return max < min ? min : max;
};

export const getStep = (value) => {
  return isInteger(value) ? value : 1;
};

export const isInteger = (value) => {
  if (typeof value === "number") {
    return true;
  } else if (typeof value === "string") {
    return value.match(NUMBERS_REG);
  } else {
    return false;
  }
};

export default RangeSlider;
