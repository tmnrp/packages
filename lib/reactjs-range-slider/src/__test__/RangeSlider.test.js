import React from "react";
import "@testing-library/jest-dom/extend-expect";
import { render, fireEvent } from "@testing-library/react";
import RangeSlider, {
  POSITION_LIST,
  getSliderThumbShapeCls,
  getLegendPositionCls,
  legendOnTopOfSliderHandler,
  onSlide,
  getLegendOnSlider,
  getLegend,
  isInteger,
  getSliderValue,
  getMin,
  getMax,
  getStep
} from "../RangeSlider";

test("Position list should be accessible", () => {
  expect(POSITION_LIST).toBeDefined();
});

test("Position list should contain values", () => {
  expect(POSITION_LIST.length > 0).toBeTruthy();
});

test(" Should return default cls with attached class with valid Param", () => {
  expect(getSliderThumbShapeCls("square")).toEqual("range-slider square");
});

describe("Should return default cls", () => {
  test("Empty param", () => {
    expect(getSliderThumbShapeCls()).toEqual("range-slider");
  });

  test("null", () => {
    expect(getSliderThumbShapeCls(null)).toEqual("range-slider");
  });

  test("undefined", () => {
    expect(getSliderThumbShapeCls(undefined)).toEqual("range-slider");
  });

  test("Empty string", () => {
    expect(getSliderThumbShapeCls("")).toEqual("range-slider");
  });
});

describe("Legends CLS should return false", () => {
  test("Empty param", () => {
    expect(getLegendPositionCls()).toEqual(false);
  });

  test("null", () => {
    expect(getLegendPositionCls(null)).toEqual(false);
  });

  test("undefined", () => {
    expect(getLegendPositionCls(undefined)).toEqual(false);
  });

  test("Empty string", () => {
    expect(getLegendPositionCls("")).toEqual(false);
  });

  test("Invalid string", () => {
    expect(getLegendPositionCls("invalid")).toEqual(false);
  });
});

describe("Legends CLS should return appopriate value", () => {
  test("top", () => {
    expect(getLegendPositionCls("top")).toEqual("legend-top");
  });
  test("right", () => {
    expect(getLegendPositionCls("right")).toEqual("legend-right");
  });
  test("bottom", () => {
    expect(getLegendPositionCls("bottom")).toEqual("legend-bottom");
  });
  test("left", () => {
    expect(getLegendPositionCls("left")).toEqual("legend-left");
  });
});

describe("DOM testing: Legend on top of slider", () => {
  const testingEl = (
    <div data-testid="myid" id="myid" className="range-slider-value">
      ABC
    </div>
  );

  test("Valid value", () => {
    const { getByTestId } = render(testingEl);
    legendOnTopOfSliderHandler(100);
    expect("130px").toEqual(getByTestId("myid").style.left);
  });
  test("Invalid value", () => {
    const { getByTestId } = render(testingEl);
    legendOnTopOfSliderHandler("abcd");
    expect("").toEqual(getByTestId("myid").style.left);
  });
  test("Empty param", () => {
    const { getByTestId } = render(testingEl);
    legendOnTopOfSliderHandler();
    expect("").toEqual(getByTestId("myid").style.left);
  });
  test("null param", () => {
    const { getByTestId } = render(testingEl);
    legendOnTopOfSliderHandler(null);
    expect("").toEqual(getByTestId("myid").style.left);
  });
  test("undefined param", () => {
    const { getByTestId } = render(testingEl);
    legendOnTopOfSliderHandler(undefined);
    expect("").toEqual(getByTestId("myid").style.left);
  });

  test("Valid value", () => {
    let val = 50;
    const testingEl = (
      <input
        data-testid="myid"
        type="range"
        value={val}
        onChange={(e) => onSlide(e.currentTarget, setValue, true)}
      />
    );
    const setValue = (v) => {
      val = v;
    };
    const { getByTestId } = render(testingEl);
    const el = getByTestId("myid");
    fireEvent.change(el, { target: { value: "100" } });
    expect(val).toEqual("100");
  });
});

test("legend on slider renders", () => {
  expect(getLegendOnSlider(50, true)).toMatchSnapshot();
});

describe("Legend container", () => {
  test("Legend function should exist", () => {
    expect(typeof getLegend).toBeDefined();
  });

  test("Valid parameters should return valid element", () => {
    const { getByTestId } = render(
      getLegend("start", "label", "75", { backgroundColor: "red" }, true)
    );
    expect(getByTestId("legend-testid")).toMatchSnapshot();
  });

  test("Empty param", () => {
    const { getByTestId } = render(getLegend(null, null, null, null, true));
    expect(getByTestId("legend-testid")).toMatchSnapshot();
  });

  test("Invalid param", () => {
    const { getByTestId } = render(
      getLegend("Invalid value", null, null, null, true)
    );
    expect(getByTestId("legend-testid")).toMatchSnapshot();
  });

  test("Falsy alignment test", () => {
    const { getByTestId } = render(getLegend(null, null, null, null, true));
    expect(getByTestId("legend-testid")).toHaveClass("legend");
  });

  test("Truthy alignment test", () => {
    const { getByTestId } = render(getLegend("end", null, null, null, true));
    expect(getByTestId("legend-testid")).toHaveClass("legend end");
  });

  test("Truthy alignment test", () => {
    expect(getLegend("end", null, null, null, false)).toBeNull();
  });
});

describe("Slider rendering", () => {
  const sliderEl = (
    <RangeSlider
      value={50}
      min={0}
      max={100}
      step={1}
      sliderThumbShape="square"
      legend={true}
      legendLabel="Range: "
      legendPosition={"top"}
      legendAlignment={"end"}
    />
  );

  test("Slider is defined", () => {
    expect(RangeSlider).toBeDefined();
  });

  test("Slider renders", () => {
    const { getByTestId } = render(sliderEl);
    expect(getByTestId("range-slider-container-testid")).toMatchSnapshot();
  });

  test("Slider renders", () => {
    const { getByTestId } = render(sliderEl);
    const sliderValue = "100";
    const el = getByTestId("range-slider-container-testid").querySelector(
      "#rangeslider"
    );
    fireEvent.change(el, {
      target: { value: sliderValue }
    });
    expect(el.value).toEqual(sliderValue);
  });

  test("Legend should be rendered", () => {
    const { getByTestId } = render(sliderEl);
    expect(getByTestId("legend-testid")).toHaveClass("legend end");
  });
});

describe("Is integer test", () => {
  test("valid integer", () => {
    expect(isInteger(1)).toBeTruthy();
  });
  test("Valid string", () => {
    expect(isInteger("1")).toBeTruthy();
  });
  test("Invalid string", () => {
    expect(isInteger("a")).toBeFalsy();
  });
  test("null", () => {
    expect(isInteger(null)).toBeFalsy();
  });
  test("undefined", () => {
    expect(isInteger(undefined)).toBeFalsy();
  });
  test("Array", () => {
    expect(isInteger([])).toBeFalsy();
  });
  test("Object", () => {
    expect(isInteger({})).toBeFalsy();
  });
});

describe("Valid integer test", () => {
  test("Valid integer value renders", () => {
    const sliderEl = <RangeSlider value={75} min={10} max={90} step={2} />;
    const { getByTestId } = render(sliderEl);
    const el = getByTestId("range-slider-container-testid").querySelector(
      "#rangeslider"
    );
    expect(el).toMatchSnapshot();
  });
  test("Valid integer value renders", () => {
    const sliderEl = (
      <RangeSlider value={null} min={null} max={null} step={null} />
    );
    const { getByTestId } = render(sliderEl);
    const el = getByTestId("range-slider-container-testid").querySelector(
      "#rangeslider"
    );
    expect(el).toMatchSnapshot();
  });

  test("Valid range value", () => {
    expect(getSliderValue(1)).toEqual(1);
  });
  test("Invalid/Default range value", () => {
    expect(getSliderValue(null)).toEqual(0);
  });

  test("Valid min", () => {
    expect(getMin(1)).toEqual(1);
  });
  test("Invalid/Default min", () => {
    expect(getMin(null)).toEqual(0);
  });

  test("Valid max", () => {
    expect(getMax(1)).toEqual(1);
  });
  test("Invalid/Default max", () => {
    expect(getMax(null)).toEqual(100);
  });
  test("valid max < min", () => {
    expect(getMax(50, 60)).toEqual(60);
  });
  test("Invalid/Default max < min", () => {
    expect(getMax(50, 0)).toEqual(50);
  });

  test("Valid step", () => {
    expect(getStep(1)).toEqual(1);
  });
  test("Invalid/Default step", () => {
    expect(getStep(null)).toEqual(1);
  });
});
