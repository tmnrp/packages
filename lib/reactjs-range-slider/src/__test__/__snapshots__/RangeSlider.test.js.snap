// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Legend container Empty param 1`] = `
<span
  class="legend"
  data-testid="legend-testid"
>
   
</span>
`;

exports[`Legend container Invalid param 1`] = `
<span
  class="legend"
  data-testid="legend-testid"
>
   
</span>
`;

exports[`Legend container Valid parameters should return valid element 1`] = `
<span
  class="legend start"
  data-testid="legend-testid"
  style="background-color: red;"
>
  label 75
</span>
`;

exports[`Slider rendering Slider renders 1`] = `
<span
  class="range-slider-container legend-top"
  data-testid="range-slider-container-testid"
>
  <span
    class="range-slider-wrap undefined"
  >
    <span
      class="min-wrap"
    >
      0
    </span>
    <span
      class="custom-range-slider"
    >
      <input
        class="range-slider square"
        id="rangeslider"
        max="100"
        min="0"
        name="rangeslider"
        step="1"
        type="range"
        value="50"
      />
    </span>
    <span
      class="max-wrap"
    >
      100
    </span>
  </span>
  <span
    class="legend end"
    data-testid="legend-testid"
  >
    Range:  50
  </span>
</span>
`;

exports[`Valid integer test Valid integer value renders 1`] = `
<input
  class="range-slider"
  id="rangeslider"
  max="90"
  min="10"
  name="rangeslider"
  step="2"
  type="range"
  value="75"
/>
`;

exports[`Valid integer test Valid integer value renders 2`] = `
<input
  class="range-slider"
  id="rangeslider"
  max="100"
  min="0"
  name="rangeslider"
  step="1"
  type="range"
  value="0"
/>
`;

exports[`legend on slider renders 1`] = `
<React.Fragment>
  <span
    className="range-slider-thumb"
  />
  <span
    className="range-slider-value"
  >
    50
  </span>
</React.Fragment>
`;
